go/kbfs/simplefs/downloads.go

```diff
@@ -0,0 +1,353 @@
+// Use of this source code is governed by a BSD
+// license that can be found in the LICENSE file.
+
+package simplefs
+
+import (
+	"fmt"
+	"os"
+	"path"
+	"path/filepath"
+	"sort"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/keybase/client/go/kbfs/libkbfs"
+	"github.com/keybase/client/go/libkb"
+	"github.com/keybase/client/go/protocol/keybase1"
+	"github.com/pkg/errors"
+	"golang.org/x/net/context"
+)
+
+const (
+	// dlCtxOpID is the display name for the unique operation SimpleFS ID tag.
+	dlCtxOpID = "SFSDLID"
+)
+
+// dlCtxTagKey is the type used for unique context tags
+type dlCtxTagKey int
+
+const (
+	// dlCtxIDKey is the type of the tag for unique operation IDs.
+	dlCtxIDKey dlCtxTagKey = iota
+)
+
+type download struct {
+	info  keybase1.DownloadInfo
+	state keybase1.DownloadState
+	opid  keybase1.OpID
+}
+
+// downloadManager manages downloads.
+type downloadManager struct {
+	k         *SimpleFS
+	publisher libkbfs.SubscriptionManagerPublisher
+
+	lock        sync.RWMutex
+	cacheDir    string
+	downloadDir string
+	downloads   map[string]download // download ID -> download
+}
+
+func newDownloadManager(simpleFS *SimpleFS) *downloadManager {
+	return &downloadManager{
+		k:           simpleFS,
+		publisher:   simpleFS.config.SubscriptionManagerPublisher(),
+		cacheDir:    simpleFS.config.KbEnv().GetCacheDir(),
+		downloadDir: filepath.Join(simpleFS.config.KbEnv().GetHome(), "Downloads"),
+		downloads:   make(map[string]download),
+	}
+}
+
+func (m *downloadManager) makeContext() (ctx context.Context, cancel func()) {
+	return context.WithCancel(libkbfs.CtxWithRandomIDReplayable(context.Background(), dlCtxIDKey, dlCtxOpID, m.k.log))
+}
+
+func (m *downloadManager) getDownload(downloadID string) (download, error) {
+	m.lock.RLock()
+	defer m.lock.RUnlock()
+	d, ok := m.downloads[downloadID]
+	if !ok {
+		return download{}, errors.New("unknown downloadID")
+	}
+	return d, nil
+}
+
+func (m *downloadManager) updateDownload(downloadID string, f func(original download) download) (err error) {
+	defer m.publisher.DownloadStatusChanged()
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	download, ok := m.downloads[downloadID]
+	if !ok {
+		return errors.New("unknown downloadID")
+	}
+	m.downloads[downloadID] = f(download)
+	return nil
+}
+
+func (m *downloadManager) monitorDownload(
+	ctx context.Context, opid keybase1.OpID, downloadID string,
+	done func(error)) {
+	ticker := time.NewTicker(time.Second)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ticker.C:
+			resp, err := m.k.SimpleFSCheck(ctx, opid)
+			switch err {
+			case nil:
+				if err := m.updateDownload(downloadID, func(d download) download {
+					d.state.EndEstimate = resp.EndEstimate
+					d.state.Progress = float64(
+						resp.BytesWritten) / float64(resp.BytesTotal)
+					return d
+				}); err != nil {
+					done(err)
+					return
+				}
+			case errNoResult:
+				// This is from simpleFS. Likely download has finished, but
+				// wait for the done ch.
```

Comment:
What does "wait for the done ch" mean?  In this code, we never see the done channel.

Reply:
I think I meant to say `ctx.Done()`, which gets canceled through `done()`. Will fix; sorry!